arch/xtensa/core/fatal.c

```diff
@@ -9,6 +9,8 @@
 #include <inttypes.h>
 #include <kernel_arch_data.h>
 #include <xtensa/specreg.h>
+#include <logging/log.h>
```

Comment:
So now every user of z_fatal_print() now has to pull in logging/log.h by themselves, and also include LOG_MODULE_DECLARE(os) otherwise they will get weird build errors.

Sorry, I'm not seeing how this is an improvement.

Reply:
well, what is wrong with including log.h? `LOG_MODULE_DECLARE` does nothing when log is disabled.

Reply:
Before, all people had to do was pull in fatal.h and then they could use z_fatal_print(). Now they have to do even more stuff. To what benefit? Why do you think this is somehow *better*?

Reply:
This is an internal code of kernel so `all people` is just few which will extend current fatal error framework in the future. A benefit is the use of log or printk directly which gives compiler (not preprocessor) awareness what can be removed. That allows code cleaning in places like:
https://github.com/zephyrproject-rtos/zephyr/blob/75bf3c5368f330a7730d7a733b56129bdfe61b92/kernel/fatal.c#L47 or https://github.com/zephyrproject-rtos/zephyr/blob/75bf3c5368f330a7730d7a733b56129bdfe61b92/arch/arm/core/cortex_m/fault.c#L25
Try disabling printk and logger on riscv and you will get warning of unused function:
https://github.com/zephyrproject-rtos/zephyr/blob/75bf3c5368f330a7730d7a733b56129bdfe61b92/arch/riscv/core/fatal.c#L34

Additionally, you get all logger features (e.g. runtime filtering), current and future one. log support (removed by #18060) was a hack. You titled your PR `temp remove log` so your intention was to bring it back in the future. 

If it was my decision, i would used log directly there and remove printk. printk does not offer anything more than logger. logger can be synchronous as well (immediate option and always in panic), has log levels (see poor version of that here: https://github.com/zephyrproject-rtos/zephyr/blob/75bf3c5368f330a7730d7a733b56129bdfe61b92/arch/arm/core/cortex_m/fault.c#L31
It supports multiple backends and eventually it will support logging to flash (crashlog). Once crashlog is in place those printk's will become a burden rather than benefit.

Reply:
> If it was my decision, i would used log directly there and remove printk. 

This is somewhere I think we can come to an agreement. The problems I am having are the different usage models for LOG and printk which is making this very complicated and ugly.

Let's consider a future where z_fatal_print() doesn't exist and we just have LOG. Can we make LOG more usable in its default configuration at least when CONFIG_TEST is enabled? Right now the buffer is very small, if you get a crash you lose lots of messages, I would rather at least for tests (which regularly find all kinds of ways to blow up) that the default is for immediate unbuffered logging so in crash or CI logs there's no chance you'll miss anything.

If you can agree to that then I'm fine with replacing all these z_fatal_print() instances with log statements.